### ISSUE TITLE
[systemd-gpt-auto-generator]: Assertion 'close_nointr(fd) != -EBADF' failed

### ISSUE THREAD

**Issue by adrian5** [systemd/systemd#25348]
### systemd version the issue has been seen with

252.1-1-arch

### Used distribution

Arch Linux

### Linux kernel version used

6.0.8-arch1-1

### CPU architectures issue was seen on

x86_64

### Component

systemd-gpt-auto-generator

### Expected behaviour you didn't see

No error messages

### Unexpected behaviour you saw

During boot I get the following message:

```
systemd-gpt-auto-generator[290]: Assertion 'close_nointr(fd) != -EBADF' failed at src/basic/fd-util.c:77, function safe_close(). Aborting.
systemd[283]: /usr/lib/systemd/system-generators/systemd-gpt-auto-generator terminated by signal ABRT.
```

**Comment by adrian5** [comment]
I think this started to occur with the most recent version update (a 252 patch) which I did yesterday.

**Comment by yuwata** [comment]
Could you provide the result of the following command?
```
mkdir /tmp/foo
sudo SYSTEMD_LOG_LEVEL=debug /usr/lib/systemd/system-generators/systemd-gpt-auto-generator /tmp/foo /tmp/foo /tmp/foo
```

**Comment by yuwata** [comment]
Does #25370 fix the issue?

**Comment by adrian5** [comment]
Here's the output to your query:

```
$ mkdir /tmp/foo
$ sudo SYSTEMD_LOG_LEVEL=debug /usr/lib/systemd/system-generators/systemd-gpt-auto-generator /tmp/foo /tmp/foo 

Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Found container virtualization none.
Disabling root partition auto-detection, root= is defined.
Failed to chase block device, ignoring: No such file or directory
Kernel was quicker than us in adding partition 1.
Opened /dev/sda1 (fd=5, whole_block_devnum=8:0, diskseq=1).
Kernel was quicker than us in adding partition 2.
Opened /dev/sda2 (fd=6, whole_block_devnum=8:0, diskseq=1).
No root partition found of the native architecture, falling back to a root partition of the secondary architecture.
Probed fstype 'ext4' on partition /dev/sda2.
/efi specified in fstab, ignoring.
Parameter ro/rw specified on kernel command line, not generating drop-in for systemd-remount-fs.service.
Assertion 'close_nointr(fd) != -EBADF' failed at src/basic/fd-util.c:77, function safe_close(). Aborting.
fish: Job 1, 'sudo SYSTEMD_LOG_LEVEL=debug /u…' terminated by signal SIGABRT (Abort)
```

**Comment by adrian5** [comment]
> Does #25370 fix the issue?

I'd need to build and run my own copy of systemd, right? This is my main workstation so I don't think I'd be comfortable attempting that. Apologies!

**Comment by yuwata** [comment]
> > Does #25370 fix the issue?
> 
> I'd need to build and run my own copy of systemd, right?

No, that's not necessary. Just cloning the repository, building the branch, and executing `systemd-gpt-auto-generator` in the build directory is enough. That should not break your system.

**Comment by adrian5** [comment]
Confirmed fixed in latest Arch release (`252.2-1-arch`):

```
…
/efi specified in fstab, ignoring.
Parameter ro/rw specified on kernel command line, not generating drop-in for systemd-remount-fs.service.

↑ last output line
```
